src/asar-utils.ts

```diff
@@ -31,3 +58,153 @@ export const generateAsarIntegrity = (asarPath: string) => {
+  x64,
+  arm64,
+  output,
+  lipo = 'lipo',
+  uniqueAllowList,
+}: FuseASARsOptions): Promise<void> => {
+  d(`fusing ${x64} and ${arm64}`);
+
+  const x64Files = new Set(asar.listPackage(x64).map(toRelativePath));
+  const arm64Files = new Set(asar.listPackage(arm64).map(toRelativePath));
+
+  //
+  // Build set of unpacked directories and files
+  //
+
+  const unpackedFiles = new Set<string>();
+
+  function buildUnpacked(a: string, fileList: Set<string>): void {
+    for (const file of fileList) {
+      const stat = asar.statFile(a, file);
+
+      if (!('unpacked' in stat) || !stat.unpacked) {
+        continue;
+      }
+
+      if ('files' in stat) {
+        continue;
+      }
+      unpackedFiles.add(file);
+    }
+  }
+
+  buildUnpacked(x64, x64Files);
+  buildUnpacked(arm64, arm64Files);
+
+  //
+  // Build list of files/directories unique to each asar
+  //
+
+  for (const file of x64Files) {
+    if (!arm64Files.has(file)) {
+      checkUnique(x64, file, uniqueAllowList);
+    }
+  }
+  const arm64Unique = [];
+  for (const file of arm64Files) {
+    if (!x64Files.has(file)) {
+      checkUnique(arm64, file, uniqueAllowList);
+      arm64Unique.push(file);
+    }
+  }
+
+  //
+  // Find common bindings with different content
+  //
+
+  const commonBindings = [];
+  for (const file of x64Files) {
+    if (!arm64Files.has(file)) {
+      continue;
+    }
+
+    // Skip directories
+    if (isDirectory(x64, file)) {
+      continue;
+    }
+
+    const x64Content = asar.extractFile(x64, file);
+    const arm64Content = asar.extractFile(arm64, file);
+
+    if (x64Content.compare(arm64Content) === 0) {
+      continue;
+    }
+
+    if (!MACHO_MAGIC.has(x64Content.readUInt32LE(0))) {
+      throw new Error(`Can't reconcile two non-macho files ${file}`);
+    }
+
+    commonBindings.push(file);
+  }
+
+  //
+  // Extract both
+  //
+
+  const x64Dir = await fs.mkdtemp(path.join(os.tmpdir(), 'x64-'));
+  const arm64Dir = await fs.mkdtemp(path.join(os.tmpdir(), 'arm64-'));
+
+  try {
+    d(`extracting ${x64} to ${x64Dir}`);
+    asar.extractAll(x64, x64Dir);
+
+    d(`extracting ${arm64} to ${arm64Dir}`);
+    asar.extractAll(arm64, arm64Dir);
+
+    for (const file of arm64Unique) {
+      const source = path.resolve(arm64Dir, file);
+      const destination = path.resolve(x64Dir, file);
+
+      if (isDirectory(arm64, file)) {
+        d(`creating unique directory: ${file}`);
+        await fs.mkdirp(destination);
+        continue;
+      }
+
+      d(`xopying unique file: ${file}`);
+      await fs.mkdirp(path.dirname(destination));
+      await fs.copy(source, destination);
+    }
+
+    for (const binding of commonBindings) {
+      const source = await fs.realpath(path.resolve(arm64Dir, binding));
+      const destination = await fs.realpath(path.resolve(x64Dir, binding));
+
+      d(`fusing binding: ${binding}`);
+      execFileSync(lipo, [source, destination, '-create', '-output', destination]);
```

Comment:
Because of the way MACHO_MAGIC works couldn't this be trying to fuse universal/x64 or x64 and x64 that don't match? Not familiar enough with `lipo` to know what happens when you try to merge things like that.

Reply:
I think I should just drop the Universal constant from `MACHO_MAGIC`. The merging of x64 and x64 would fail AFAIK and it should be fine because the binding binaries shouldn't be different unless they were compiled for different platforms!